keps/sig-storage/4958-csi-sidecars-all-in-one/README.md

```diff
@@ -0,0 +1,1141 @@
+-->
+
+![overview](./aio6.png "overview of design details")
+
+#### Glossary
+- Individual repository - An existing repository in the kubernetes-csi/ org in Github e.g. the external-attacher repository.
+- AIO monorepo or monorepo - The monolithic repository where most of the code of the CSI Sidecars will be migrated.
+- Monorepo component - The source code of an individual repository that is currently being migrated or already migrated to the monorepo. 
+
+#### AIO Monorepo 
+
+##### Release Management
+
+Semantic Versioning
+
+##### RBAC policy
+
+We designed the AIO repo's RBAC policy to mirror that of individual repos, where each controller maintains its own policy. Driver maintainers should apply proper RBAC when enabling specific controllers in AIO
+
+> We plan to combine informer caches of different controllers in the future, which will involve updating some RBAC policies, to be addressed in a subsequent KEP
+
+##### Command Line
+
+Divided the command lines into two types, a generic command line whose configuration is common to all controllers and is configured only once, and the other type of command lines whose configuration is different for each controller. these command lines each has a new unique name. prefix with the controller name.
+
+example PR: https://github.com/kubernetes-csi/external-attacher/pull/620
+
+#### Monorepo component
+
+poc version: https://github.com/mauriciopoppe/csi-sidecars
+
+
+#### Deployment Workflow Definition
+
+After we see the Monorepo component running fine in integration/e2e tests in k8s, we need to perform a hard cut so that new deployment goes in the monorepo component only.
+
+##### AIO MonoRepo state definition
+
+- inner-verified: Current state of AIO MonoRepo
+- production-verified: Available for production environments, a Cloud vendor that's using it in its production environment.
+- released: Official released, Available for accept PRs from SIG Storage Developer
+- standalone: Never need sync codes from individual repos, AIO MonoRepo become the source of truth
+
+
+##### Individual repository state definition
+
+- Released: current state of individual repos
+- FeatureFreeze: 
+    - Any new feature PRs are not allowed to be filed to the master branch or release-X branches(Controlled by the individual repo maintainer, categorize it and reject it if it's a feature)
+    - SIG Storage Developer file the feature PRs to AIO MonoRepo 
+    - Except for the serious bugfixes or CVE fixes PRs (only from individual repo maintainer) which can be merged in master and backported to the other release-X branches
+- Deprecated:
+    - Not maintaining this repository
+    - Eventually the image is going away for the individual repo is going away (although wouldn't possible unless we migrate ALL the sidecars)
+    - (future) archive it but not at the same time as the deprecation time, this is a terminal state so we can't undo it
+
+
+![state change](./aio11.png "overview of workflow definition")
```

Comment:
Feature freeze on the individual repos should start after the monorepo reaches GA.